ouroboros-network/sim-tests-lib/Test/Ouroboros/Network/PeerSelection.hs

```diff
@@ -1511,14 +1546,57 @@ recentPeerShareActivity d =
         (E t (GovernorEvent (TraceDemoteBigLedgerPeersAsynchronous m)) : txs) =
       let peersDemotedToCold = Map.foldrWithKey'
                                 (\k v r -> case v of
-                                  (PeerCold, _) -> k : r
-                                  _             -> r
+                                  (PeerCold, _)    -> k : r
+                                  (PeerCooling, _) -> k : r
+                                  _                -> r
                                 ) [] m
           recentSet' = foldl' (flip Set.delete) recentSet peersDemotedToCold
           recentPSQ' = foldl' (flip PSQ.delete) recentPSQ peersDemotedToCold
        in E t (Nothing, recentSet')
         : go recentSet' recentPSQ' txs
 
+    go !recentSet !recentPSQ
+        (E t (GovernorEvent (TracePromoteWarmBigLedgerPeerFailed _ _ addr _)) : txs) =
+      let recentSet' = Set.delete addr recentSet
+          recentPSQ' = PSQ.delete addr recentPSQ
+       in E t (Nothing, recentSet')
+        : go recentSet' recentPSQ' txs
+
+    go !recentSet !recentPSQ
+        (E t (GovernorEvent (TracePromoteWarmFailed _ _ addr _)) : txs) =
+      let recentSet' = Set.delete addr recentSet
+          recentPSQ' = PSQ.delete addr recentPSQ
+       in E t (Nothing, recentSet')
+        : go recentSet' recentPSQ' txs
+
+    go !recentSet !recentPSQ
+        (E t (GovernorEvent (TraceDemoteHotFailed _ _ addr _)) : txs) =
+      let recentSet' = Set.delete addr recentSet
+          recentPSQ' = PSQ.delete addr recentPSQ
+       in E t (Nothing, recentSet')
+        : go recentSet' recentPSQ' txs
+
+    go !recentSet !recentPSQ
+        (E t (GovernorEvent (TraceDemoteWarmFailed _ _ addr _)) : txs) =
+      let recentSet' = Set.delete addr recentSet
+          recentPSQ' = PSQ.delete addr recentPSQ
+       in E t (Nothing, recentSet')
+        : go recentSet' recentPSQ' txs
+
+    go !recentSet !recentPSQ
+        (E t (GovernorEvent (TraceDemoteHotBigLedgerPeerFailed _ _ addr _)) : txs) =
+      let recentSet' = Set.delete addr recentSet
+          recentPSQ' = PSQ.delete addr recentPSQ
+       in E t (Nothing, recentSet')
+        : go recentSet' recentPSQ' txs
+
+    go !recentSet !recentPSQ
+        (E t (GovernorEvent (TraceDemoteWarmBigLedgerPeerFailed _ _ addr _)) : txs) =
+      let recentSet' = Set.delete addr recentSet
+          recentPSQ' = PSQ.delete addr recentPSQ
+       in E t (Nothing, recentSet')
+        : go recentSet' recentPSQ' txs
+
```

Comment:
Could you add haddocks to the definition of `recentSet` and `recentPSQ` where `go` type signature is provided.  Do I understand correctly that these are set of peers available for peer sharing?

Btw, do we need both `recentSet` and `recentPSQ` if we always update both in the same way?

Reply:
It was @dcoutts that wrote this test, but my guess is that we need both sets information in order to check that we don't share with the same peer too often.

Reply:
Ask Duncan and let's add the haddocks.

Reply:
I have added some of what I think are the sets. They currently mimic the `nextPeerShareTimes` set in `EstablishedPeers` type and then look at the current/recent set of peer shared peers and check that the intersection between the two is empty.